src/frontegg-auth/src/auth.rs

```diff
@@ -117,115 +304,225 @@ impl Authentication {
-                        match res.await {
-                            Ok(res) => return res,
-                            Err(e) => {
-                                // Log at info level, not warn or error level,
-                                // because the error could be a result of a
-                                // user's permissions being intentionally
-                                // revoked.
-                                //
-                                // TODO: report Frontegg errors at `error!`
-                                // level or via a Prometheus metric so that
-                                // we can alert on them.
-                                info!(
-                                    "failed to refresh token for {}: {:#}",
-                                    response.claims.email, e
-                                );
-                                // Retry again later. 5 seconds chosen arbitrarily.
-                                time::sleep(Duration::from_secs(5)).await;
-                            }
-                        }
-                    }
-                };
-                let expire_in =
-                    u64::try_from(response.claims.exp - frontegg.now.as_secs()).unwrap_or(0);
-                let expire_in = time::sleep(Duration::from_secs(expire_in));
-
-                tokio::select! {
-                    _ = expire_in => return,
-                    res = refresh_request => {
-                        response = res;
-                        claims_processor(&response.claims);
-                    },
-                };
+/// Returns a [`Duration`] for how long the provided claims are valid for.
+fn valid_for(now: &NowFn, claims: &ValidatedClaims) -> Duration {
+    let valid_for = claims.exp - now.as_secs();
+    let valid_for = u64::try_from(valid_for).unwrap_or(0);
+
+    Duration::from_secs(valid_for)
+}
+
+/// Returns a `Future` that resolves then the sending side of the provided channel closes.
+fn make_valid_until_future(
+    refresh_rx: &watch::Receiver<ExternalUserMetadata>,
+) -> BoxFuture<'static, ()> {
+    // Our auth is valid for as long as the sender is alive.
+    let mut refresh_rx = refresh_rx.clone();
+    let future = async move { while let Ok(_) = refresh_rx.changed().await {} };
+    Box::pin(future)
+}
+
+/// Spawns a task that will continuously refresh the `RefreshToken` provided in `first_response`.
+fn continuously_refresh(
+    admin_api_token_url: &str,
+    client: Client,
+    metrics: Metrics,
+    first_response: ApiTokenResponse,
+    expected_email: String,
+    validation_config: ValidationConfig,
+    refresh_tx: watch::Sender<ExternalUserMetadata>,
+) -> RefreshTaskId {
+    let refresh_url = format!("{}{}", admin_api_token_url, REFRESH_SUFFIX);
+    let id = RefreshTaskId::new();
+
+    // Continuously refresh.
+    let name = format!("frontegg-auth-refresh-{}", id);
+    mz_ore::task::spawn(|| name, async move {
```

Comment:
the nested spawn is a bit confusing but nothing seems wrong with it

Reply:
Yeah I'm not the biggest fan of it, but I don't think there is a way to avoid it in the current setup